storage/cache/subtree_cache.go

```diff
@@ -0,0 +1,269 @@
+	c := s.subtrees[prefixKey]
+	if c == nil {
+		// Cache miss, so we'll try to fetch from storage.
+		subID := id
+		subID.PrefixLenBits = len(px) * 8
+		var err error
+		c, err = getSubtree(subID)
+		if err != nil {
+			return nil, err
+		}
+		if c == nil {
+			// storage didn't have one for us, so we'll store an empty proto here
+			// incase we try to update it later on (we won't flush it back to
+			// storage unless it's been written to.)
+			c = &storage.SubtreeProto{
+				Prefix: px,
+				Depth:  strataDepth,
+				Leaves: make(map[string][]byte),
+			}
+		}
+		// Whether from storage or freshly created, we'll need an internal nodes cache map:
+		c.InternalNodes = make(map[string][]byte)
+		if c.Prefix == nil {
+			panic(fmt.Errorf("GetNodeHash nil prefix on %v for id %v with px %#v", c, id.String(), px))
+		}
+
+		s.subtrees[prefixKey] = c
+	}
+
+	// finally look for the particular node within the subtree so we can return
+	// the hash & revision.
+	var nh trillian.Hash
+	if sx.bits == 8 {
+		nh = c.Leaves[sx.serialize()]
+	} else {
+		nh = c.InternalNodes[sx.serialize()]
+	}
+	if nh == nil {
+		return nil, nil
+	}
+	return nh, nil
+}
+
+// SetNodeHash sets a node hash in the cache.
+func (s *SubtreeCache) SetNodeHash(id storage.NodeID, h trillian.Hash) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	px, sx := splitNodeID(id)
+	prefixKey := string(px)
+	c := s.subtrees[prefixKey]
+	if c == nil {
+		// TODO(al): This is ok, IFF *all* leaves in the subtree are being set,
+		// verify that this is the case when it happens.
+		c = &storage.SubtreeProto{
+			Prefix:        px,
+			Leaves:        make(map[string][]byte),
+			InternalNodes: make(map[string][]byte),
+		}
+		s.subtrees[prefixKey] = c
+	}
+	if c.Prefix == nil {
+		panic(fmt.Errorf("nil prefix for %v (key %v)", id.String(), prefixKey))
+	}
+	s.dirtyPrefixes[prefixKey] = true
+	if sx.bits == 8 {
+		c.Leaves[sx.serialize()] = h
+	} else {
+		c.InternalNodes[sx.serialize()] = h
+	}
+	return nil
+}
+
+// Flush causes the cache to write all dirty Subtrees back to storage.
+func (s *SubtreeCache) Flush(setSubtree SetSubtreeFunc) error {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	for k, v := range s.subtrees {
+		if s.dirtyPrefixes[k] {
+			bk := []byte(k)
+			if !bytes.Equal(bk, v.Prefix) {
+				return fmt.Errorf("inconsistent cache: prefix key is %v, but cached object claims %v", bk, v.Prefix)
+			}
+			if len(v.Leaves) > 0 {
+				// clear the internal node cache; we don't want to write that.
+				v.InternalNodes = nil
+				if err := setSubtree(v); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// makeSuffix creates a suffix key for indexing into the subtree's Leaves and
+// InternalNodes maps.
+func makeSuffixKey(depth int, index int64) (string, error) {
+	// TODO(al): only supports 8 bit subtree sizes currently
+	if depth >= 8 || depth < 0 {
+		return "", fmt.Errorf("found depth of %d, but we only support positive depths of 8 currently", depth)
```

Comment:
error message still reads like 8 is only supported value. Suggest replacing "of" with "up to"